tests/trainer/test_trainer.py

```diff
@@ -1,247 +1,375 @@
+            pytest.importorskip("torch", minversion="1.10", reason="Pytorch 1.10 required.")
+
+        # create the algorithms
+        if name in ('cutmix, mixup'):  # these algos have required algorithms
+            algorithm = hparams(num_classes=2).initialize_object()
+        else:
+            algorithm = hparams().initialize_object()
+
+        return algorithm
+
+    @pytest.fixture(params=callback_registry.items(), ids=callback_registry.keys())
+    def callback(self, request, tmpdir, monkeypatch):
+        name, hparams = request.param
+
+        if name == 'benchmarker':
+            pytest.skip('Deprecated callback.')
+
+        # create callback
+        if name == 'run_directory_uploader':
+            pytest.importorskip('libcloud', reason='libcloud required.')
+            monkeypatch.setenv("KEY_ENVIRON", str(tmpdir))
+
+            callback = hparams(
+                provider='local',
+                container='.',
+                key_environ="KEY_ENVIRON",
+            ).initialize_object()
+        else:
+            callback = hparams().initialize_object()
+
+        return callback
+
+    @pytest.fixture(params=logger_registry.items(), ids=logger_registry.keys())
+    def logger(self, request):
+
+        name, hparams = request.param
+        skip_list = {
+            'wandb': 'Requires wandb account.',
+            'mosaicml': 'Not supported',
+        }
+        if name in skip_list:
+            pytest.skip(skip_list[name])
+
+        return hparams().initialize_object()
+
+    """
+    Tests that training completes.
+    """
+
+    def test_algorithms(self, config, algorithm):
+        config['algorithms'] = [algorithm]
+        trainer = Trainer(**config)
+        trainer.fit()
+
+    @pytest.mark.timeout(10)
+    def test_callbacks(self, config, callback):
+        config['callbacks'] = [callback]
+        trainer = Trainer(**config)
+        trainer.fit()
+
+    def test_loggers(self, config, logger):
+        config['loggers'] = [logger]
+        trainer = Trainer(**config)
+        trainer.fit()
+
+    """
+    Tests that training with multiple fits complete.
+    Note: future functional tests should test for
+    idempotency (e.g functionally)
+    """
+
+    def test_algorithms_multiple_calls(self, config, algorithm):
+        if isinstance(algorithm, LayerFreezing):
+            pytest.xfail("Known idempotency issue.")
+        config['algorithms'] = [algorithm]
+        trainer = Trainer(**config)
+        self._test_multiple_fits(trainer)
+
+    def test_callbacks_multiple_calls(self, config, callback):
+        if isinstance(callback, RunDirectoryUploader):
+            pytest.xfail("Known idempotency issue.")
+        config['callbacks'] = [callback]
+        trainer = Trainer(**config)
+        self._test_multiple_fits(trainer)
+
+    def test_loggers_multiple_calls(self, config, logger):
+        if isinstance(logger, FileLogger):
+            pytest.xfail("Known idempotency issue.")
```

Comment:
It's not an issue with the logger, but rather, that the trainer automatically closes all callbacks after the first call to .fit. This can be fixed by using the private method `trainer._train_loop()` on line 373. Would rather have the file logger tested and call a private method than expect it to fail.

Reply:
See comment above. Since we already test the `FileLogger` within the single call to fit, and multiple calls to fit is not an officially supported feature, prefer to keep the test code to be the expected public usage in the future.